{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import natsort\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "\n",
    "sys.path.append(\n",
    "    \"/nemo/lab/znamenskiyp/home/users/becalia/code/multi_padlock_design/lib\"\n",
    ")\n",
    "sys.path.append(\"/nemo/lab/znamenskiyp/home/users/becalia/code/multi_padlock_design\")\n",
    "sys.path.append(\n",
    "    \"/nemo/lab/znamenskiyp/home/users/becalia/code/multi_padlock_design/notebooks\"\n",
    ")\n",
    "from lib import check_padlocks\n",
    "import config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tqdm.pandas()\n",
    "blast_file_dir = Path(\"/nemo/lab/znamenskiyp/scratch/olfr_queries\")\n",
    "ref_path = Path(\"/nemo/lab/znamenskiyp/home/shared/resources/refseq/\")\n",
    "check_padlocks.loaddb(\"mouse\", config)\n",
    "\n",
    "genes = [\n",
    "    os.path.basename(file).split(\"_query\")[0]\n",
    "    for file in os.listdir(blast_file_dir)\n",
    "    if file.endswith(\"query_blast.out\")\n",
    "]\n",
    "print(genes)\n",
    "print(len(genes))\n",
    "for gene in genes:\n",
    "    check_padlocks.find_off_targets(gene, blast_file_dir, ref_path)\n",
    "\n",
    "# make a df from all the files that end with _off_targets.out\n",
    "files = glob.glob(str(blast_file_dir / \"*_off_targets.out\"))\n",
    "# ignore the first row of each file\n",
    "df = pd.concat([pd.read_csv(file, header=0) for file in files])\n",
    "# sort df with natsort on column\n",
    "df = df.iloc[natsort.index_natsorted(df[\"0\"])]\n",
    "header = [\n",
    "    \"query\",\n",
    "    \"subject\",\n",
    "    \"percentage identity\",\n",
    "    \"length\",\n",
    "    \"mismatches\",\n",
    "    \"gaps\",\n",
    "    \"qstart\",\n",
    "    \"qend\",\n",
    "    \"sstart\",\n",
    "    \"send\",\n",
    "    \"evalue\",\n",
    "    \"bitscore\",\n",
    "    \"qseq\",\n",
    "    \"sseq\",\n",
    "    \"homology_candidate_hit\",\n",
    "    \"gene\",\n",
    "]\n",
    "# remove the first column\n",
    "df = df.drop(columns=[\"Unnamed: 0\"])\n",
    "df.columns = header\n",
    "df.sort_values(by=[\"evalue\"], inplace=True)\n",
    "\n",
    "precomputed_variants = check_padlocks.precompute_variants(df, ref_path)\n",
    "\n",
    "# Running the optimized processing function\n",
    "df = check_padlocks.process_dataframe(\n",
    "    df, armlength=20, tm_threshold=37, precomputed_variants=precomputed_variants\n",
    ")\n",
    "\n",
    "# Extract the variants\n",
    "df[\"variant\"] = df[\"subject\"].str.split(\".\", n=1, expand=True)[0]\n",
    "\n",
    "# Get unique variants to reduce the number of lookups\n",
    "unique_variants = df[\"variant\"].unique()\n",
    "\n",
    "# Get genes for unique variants\n",
    "genes_dict = check_padlocks.find_genes_from_variants(unique_variants, \"mouse\", config)\n",
    "\n",
    "# Map genes back to the DataFrame\n",
    "df[\"blast_target\"] = df[\"variant\"].map(genes_dict)\n",
    "\n",
    "# Simplify the 'offtarget' field\n",
    "df[\"blast_target\"] = df[\"blast_target\"].apply(\n",
    "    lambda x: x[0] if isinstance(x, list) and len(x) > 0 else None\n",
    ")\n",
    "\n",
    "# Drop the temporary 'variant' column\n",
    "df = df.drop(columns=[\"variant\"])\n",
    "\n",
    "# Load the conversion table\n",
    "conversion_table = pd.read_csv(Path().cwd().parent / \"data/updated_idmap.csv\")\n",
    "\n",
    "# Remove the rows that have NaN in the alias column\n",
    "conversion_table = conversion_table.dropna(subset=[\"alias\"])\n",
    "\n",
    "# Create a dictionary for fast lookups\n",
    "alias_to_symbol = {}\n",
    "\n",
    "for index, row in conversion_table.iterrows():\n",
    "    aliases = row[\"symbol\"].split(\", \")\n",
    "    for alias in aliases:\n",
    "        alias_to_symbol[alias] = row[\"query\"]\n",
    "\n",
    "\n",
    "# Function to find the symbol using the dictionary\n",
    "def find_symbol(gene_name, alias_to_symbol):\n",
    "    return alias_to_symbol.get(gene_name, gene_name)\n",
    "\n",
    "\n",
    "df[\"converted_blast_target\"] = df[\"blast_target\"].apply(\n",
    "    find_symbol, args=(alias_to_symbol,)\n",
    ")\n",
    "df[\"converted_gene_name\"] = df[\"gene\"].apply(find_symbol, args=(alias_to_symbol,))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate Melting temp (requires Melting 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "melting_5 = True\n",
    "if melting_5:\n",
    "    df = check_padlocks.process_dataframe_in_batches(df, batch_size=300)\n",
    "else:\n",
    "    df = pd.read_csv(\"blast_results_olfr_full_across_tms.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Tm vs no. padlocks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sum the True values for each \"valid_xx\" column\n",
    "valid_columns = [f\"valid_{i}\" for i in range(20, 51)]\n",
    "true_sums = df[valid_columns].sum()\n",
    "\n",
    "# Plotting the sums\n",
    "plt.figure(figsize=(8, 5), dpi=200)\n",
    "plt.plot(valid_columns, true_sums, marker=\"o\")\n",
    "\n",
    "plt.xlabel(\"Melting5 Tm\")\n",
    "plt.ylabel(\"Number of valid padlocks\")\n",
    "# plt.grid(True)\n",
    "plt.xticks(range(len(valid_columns)), range(20, 51))\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot number of valid paldocks per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped_queries = df.groupby(\"query\")\n",
    "\n",
    "df_padlocks = grouped_queries.agg(\n",
    "    valid_specific=(\n",
    "        \"query\",\n",
    "        lambda x: any(\n",
    "            df.loc[x.index, \"valid_probe_melting\"] & df.loc[x.index, \"specific\"]\n",
    "        ),\n",
    "    ),\n",
    "    valid_non_specific=(\n",
    "        \"query\",\n",
    "        lambda x: any(\n",
    "            df.loc[x.index, \"valid_probe_melting\"] & ~df.loc[x.index, \"specific\"]\n",
    "        ),\n",
    "    ),\n",
    ").reset_index()\n",
    "\n",
    "# Merge df_padlocks with the original df to bring padlock_target_gene into df_padlocks\n",
    "df_merged = pd.merge(\n",
    "    df_padlocks, df[[\"query\", \"gene\"]], on=\"query\", how=\"left\"\n",
    ").drop_duplicates()\n",
    "\n",
    "# Group by gene and aggregate valid_specific and valid_non_specific\n",
    "df_grouped_by_gene = (\n",
    "    df_merged.groupby(\"gene\")\n",
    "    .agg(\n",
    "        valid_specific=(\"valid_specific\", \"any\"),\n",
    "        valid_non_specific=(\"valid_non_specific\", \"any\"),\n",
    "    )\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Count the number of padlocks\n",
    "df_counts = (\n",
    "    df_merged.groupby(\"gene\")\n",
    "    .agg(\n",
    "        number_of_specific_padlocks=(\"valid_specific\", lambda x: x.sum()),\n",
    "        number_of_non_specific_padlocks=(\"valid_non_specific\", lambda x: x.sum()),\n",
    "    )\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "df_padlocks.to_csv(\"padlock_specificity_olfrs.csv\", index=False)\n",
    "df_grouped_by_gene.to_csv(\"gene_specificity_olfrs.csv\", index=False)\n",
    "df_counts.to_csv(\"number_of_specific_padlocks_olfrs.csv\", index=False)\n",
    "\n",
    "# Using df_counts to get the counts of specific and non-specific padlocks\n",
    "padlocks_per_gene_specific = df_counts.set_index(\"gene\")[\"number_of_specific_padlocks\"]\n",
    "padlocks_per_gene_non_specific = df_counts.set_index(\"gene\")[\n",
    "    \"number_of_non_specific_padlocks\"\n",
    "]\n",
    "\n",
    "# Combine the counts into one DataFrame for plotting\n",
    "genes = padlocks_per_gene_specific.index.union(padlocks_per_gene_non_specific.index)\n",
    "combined_counts = pd.DataFrame(\n",
    "    {\n",
    "        \"Specific Padlocks\": padlocks_per_gene_specific.reindex(genes, fill_value=0),\n",
    "        \"Non-Specific Padlocks\": padlocks_per_gene_non_specific.reindex(\n",
    "            genes, fill_value=0\n",
    "        ),\n",
    "    }\n",
    ")\n",
    "\n",
    "# Plotting the data\n",
    "fig, ax = plt.subplots(figsize=(60, 6))\n",
    "combined_counts[\"Specific Padlocks\"].plot(\n",
    "    kind=\"bar\", color=\"black\", ax=ax, position=0, width=0.4\n",
    ")\n",
    "combined_counts[\"Non-Specific Padlocks\"].plot(\n",
    "    kind=\"bar\", color=\"red\", ax=ax, position=1, width=0.4\n",
    ")\n",
    "\n",
    "# Customizing the plot\n",
    "ax.set_title(\"Specific and Non-Specific Padlocks per Gene\")\n",
    "ax.set_xlabel(\"Gene\")\n",
    "ax.set_ylabel(\"Number of Padlocks\")\n",
    "ax.legend([\"Specific Padlocks\", \"Non-Specific Padlocks\"])\n",
    "plt.xticks(rotation=90)\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot Tm from each method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 8), dpi=200)\n",
    "plt.title(\"Including ligation site missmatch seqs\")\n",
    "plt.scatter(\n",
    "    df[\"tm_left_NN\"], df[\"tm_left_melting\"], c=(df[\"mismatches\"]), s=0.1, vmin=0, vmax=6\n",
    ")\n",
    "plt.scatter(\n",
    "    df[\"tm_right_NN\"],\n",
    "    df[\"tm_right_melting\"],\n",
    "    c=(df[\"mismatches\"]),\n",
    "    s=0.1,\n",
    "    vmin=0,\n",
    "    vmax=6,\n",
    ")\n",
    "cbar = plt.colorbar(label=\"No. total mismatches/gaps\", fraction=0.046, pad=0.04)\n",
    "# Plot the line y=x\n",
    "plt.plot([-50, 100], [-50, 100], color=\"red\", lw=1, alpha=0.3)\n",
    "plt.xlabel(\"Tm_NN (C)\")\n",
    "plt.ylabel(\"Tm Melting5 (C)\")\n",
    "plt.gca().set_aspect(\"equal\", adjustable=\"box\")\n",
    "# add grid lines every 10 degrees\n",
    "plt.xticks(np.arange(-50, 110, 10))\n",
    "plt.yticks(np.arange(-50, 110, 10))\n",
    "# Add bold line at 0\n",
    "plt.axvline(0, color=\"black\", lw=0.5)\n",
    "plt.axhline(0, color=\"black\", lw=0.5)\n",
    "plt.ylim(-50, 80)\n",
    "plt.xlim(-50, 70)\n",
    "plt.grid(True, which=\"both\", linestyle=\"--\", lw=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the data\n",
    "plt.figure(figsize=(8, 8), dpi=200)\n",
    "plt.title(\"NN probe cutoff\")\n",
    "\n",
    "# Plot where valid_probe_NN is False (black)\n",
    "valid_false = df[df[\"valid_probe_NN\"] == False]\n",
    "plt.scatter(\n",
    "    valid_false[\"tm_left_NN\"],\n",
    "    valid_false[\"tm_left_melting\"],\n",
    "    c=\"black\",\n",
    "    s=0.1,\n",
    "    label=\"Valid Probe NN (False)\",\n",
    ")\n",
    "plt.scatter(\n",
    "    valid_false[\"tm_right_NN\"], valid_false[\"tm_right_melting\"], c=\"black\", s=0.1\n",
    ")\n",
    "\n",
    "# Plot where valid_probe_NN is True (green)\n",
    "valid_true = df[df[\"valid_probe_NN\"] == True]\n",
    "plt.scatter(\n",
    "    valid_true[\"tm_left_NN\"],\n",
    "    valid_true[\"tm_left_melting\"],\n",
    "    c=\"lime\",\n",
    "    s=2,\n",
    "    label=\"Valid Probe NN (True)\",\n",
    ")\n",
    "plt.scatter(\n",
    "    valid_true[\"tm_right_NN\"],\n",
    "    valid_true[\"tm_right_melting\"],\n",
    "    c=\"lime\",\n",
    "    s=2,\n",
    ")\n",
    "\n",
    "# Plot the line y=x\n",
    "plt.plot([-50, 100], [-50, 100], color=\"red\", lw=1, alpha=0.3)\n",
    "\n",
    "# Add red vertical line at 37 degrees\n",
    "plt.axvline(37, color=\"red\", lw=0.5)\n",
    "\n",
    "\n",
    "# Set labels and limits\n",
    "plt.xlabel(\"Tm_NN (C)\")\n",
    "plt.ylabel(\"Tm Melting5 (C)\")\n",
    "plt.gca().set_aspect(\"equal\", adjustable=\"box\")\n",
    "\n",
    "# Add grid lines every 10 degrees\n",
    "plt.xticks(np.arange(-50, 110, 10))\n",
    "plt.yticks(np.arange(-50, 110, 10))\n",
    "\n",
    "# Add bold line at 0\n",
    "plt.axvline(0, color=\"black\", lw=0.5)\n",
    "plt.axhline(0, color=\"black\", lw=0.5)\n",
    "\n",
    "# Set plot limits\n",
    "plt.ylim(-50, 80)\n",
    "plt.xlim(-50, 70)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(True, which=\"both\", linestyle=\"--\", lw=0.5)\n",
    "\n",
    "# Show legend\n",
    "plt.legend()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the data\n",
    "plt.figure(figsize=(8, 8), dpi=200)\n",
    "plt.title(\"Melting5 probe cutoff\")\n",
    "\n",
    "# Plot where valid_probe_NN is False (black)\n",
    "valid_false = df[df[\"valid_probe_melting\"] == False]\n",
    "plt.scatter(\n",
    "    valid_false[\"tm_left_NN\"],\n",
    "    valid_false[\"tm_left_melting\"],\n",
    "    c=\"black\",\n",
    "    s=0.1,\n",
    "    label=\"Valid Probe Melting5 (False)\",\n",
    ")\n",
    "plt.scatter(\n",
    "    valid_false[\"tm_right_NN\"], valid_false[\"tm_right_melting\"], c=\"black\", s=0.1\n",
    ")\n",
    "\n",
    "# Plot where valid_probe_NN is True (green)\n",
    "valid_true = df[df[\"valid_probe_melting\"] == True]\n",
    "plt.scatter(\n",
    "    valid_true[\"tm_left_NN\"],\n",
    "    valid_true[\"tm_left_melting\"],\n",
    "    c=\"lime\",\n",
    "    s=2,\n",
    "    label=\"Valid Probe Melting5 (True)\",\n",
    ")\n",
    "plt.scatter(valid_true[\"tm_right_NN\"], valid_true[\"tm_right_melting\"], c=\"lime\", s=2)\n",
    "\n",
    "# Add red line horizontal at 37\n",
    "plt.axhline(37, color=\"red\", lw=0.5)\n",
    "\n",
    "# Plot the line y=x\n",
    "plt.plot([-50, 100], [-50, 100], color=\"red\", lw=1, alpha=0.3)\n",
    "\n",
    "# Set labels and limits\n",
    "plt.xlabel(\"Tm_NN (C)\")\n",
    "plt.ylabel(\"Tm Melting5 (C)\")\n",
    "plt.gca().set_aspect(\"equal\", adjustable=\"box\")\n",
    "\n",
    "# Add grid lines every 10 degrees\n",
    "plt.xticks(np.arange(-50, 110, 10))\n",
    "plt.yticks(np.arange(-50, 110, 10))\n",
    "\n",
    "# Add bold line at 0\n",
    "plt.axvline(0, color=\"black\", lw=0.5)\n",
    "plt.axhline(0, color=\"black\", lw=0.5)\n",
    "\n",
    "# Set plot limits\n",
    "plt.ylim(-50, 80)\n",
    "plt.xlim(-50, 70)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(True, which=\"both\", linestyle=\"--\", lw=0.5)\n",
    "\n",
    "# Show legend\n",
    "plt.legend()\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn3\n",
    "from matplotlib_venn.layout.venn3 import DefaultLayoutAlgorithm\n",
    "\n",
    "# Create sets for each condition\n",
    "set_NN = set(df.index[df[\"valid_probe_NN\"]])\n",
    "set_melting = set(df.index[df[\"valid_probe_melting\"]])\n",
    "set_old_filters = set(df.index[df[\"valid_probe_old_filters\"]])\n",
    "\n",
    "# Create Venn Diagram\n",
    "plt.figure(figsize=(8, 8), dpi=200)\n",
    "venn3(\n",
    "    [set_NN, set_melting, set_old_filters],\n",
    "    set_labels=(\"NN\", \"Melting\", \"Old Filters\"),\n",
    "    layout_algorithm=DefaultLayoutAlgorithm(normalize_to=1),\n",
    ")\n",
    "\n",
    "plt.title(\"Venn Diagram of Valid Probes\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
